File: Чередников/hard 1932/1932.rs
use std::cell::RefCell;
use std::collections::{HashMap, VecDeque};
use std::rc::Rc;

const INF: i32 = 1_000_000;
impl Solution {
    pub fn can_merge(trees: Vec<Option<Rc<RefCell<TreeNode>>>>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut roots = HashMap::new();
        let mut counts = HashMap::new();

        for i in &trees {
            let node = i.clone().unwrap();

            let node = Rc::clone(&node);
            let node2 = Rc::clone(&node);
            roots.insert(node.borrow().val, node2);

            if let Some(el) = &node.borrow().left {
                counts.insert(el.borrow().val, 1);
            };
            if let Some(el) = &node.borrow().right {
                counts.insert(el.borrow().val, 1);
            };
        }

        let mut root = None;

        for i in trees {
            let i = i.unwrap();
            if counts.get(&i.borrow().val).is_none() {
                match root {
                    None => root = Some(i),
                    Some(_) => {
                        println!("4");
                        return None;
                    }
                }
            }
        }

        if root.is_none() {
            return None;
        }

        let mut queue = VecDeque::new();
        queue.push_back((root.clone().unwrap(), -INF, INF));
        while queue.is_empty() == false {
            let (node, min, max) = queue.pop_front().unwrap();
            if !(min < node.borrow().val && node.borrow().val < max) {
                return None;
            }

            let mut node = node.borrow_mut();

            if let Some(el) = &node.left {
                let el = Rc::clone(el);
                if let Some(root) = roots.get(&el.borrow().val) {
                    node.left = Some(Rc::clone(root));
                    roots.remove(&el.borrow().val); // !!!!!!!!!!!!!!!!
                };
                queue.push_back((node.left.clone().unwrap(), min, node.val));
            };

            if let Some(el) = &node.right {
                let el = Rc::clone(el);
                if let Some(root) = roots.get(&el.borrow().val) {
                    node.right = Some(Rc::clone(root));
                    roots.remove(&el.borrow().val); // !!!!!!!!!!!!!!!!
                };
                queue.push_back((node.right.clone().unwrap(), node.val, max));
            };
        }

        if roots.len() == 1 {
            root
        } else {
            None
        }
    }
}

File: Kvasha/1932.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode[]} trees
 * @return {TreeNode}
 */
let canMerge = function (trees) {
  let nodesFreq = {};
  let roots = new Map();
  let root;
  let queue = new Array();

  for (let tree of trees) {
    roots.set(tree.val, tree);
    if (tree.left) {
      if (!nodesFreq[tree.left.val]) nodesFreq[tree.left.val] = 0;
      nodesFreq[tree.left.val]++;
    }

    if (tree.right) {
      if (!nodesFreq[tree.right.val]) nodesFreq[tree.right.val] = 0;
      nodesFreq[tree.right.val]++;
    }
  }

  for (let tree of trees) {
    if (!nodesFreq[tree.val]) {
      if (root) return null;
      root = tree;
    }
  }

  if (!root) return null;
  queue.push([root, -Infinity, Infinity]);

  while (queue.length) {
    let [node, min, max] = queue.shift();
    let left = node.left;
    let right = node.right;

    if (node.val <= min || node.val >= max) return null;

    if (left) {
      if (roots.has(left.val)) {
        node.left = roots.get(left.val);
        roots.delete(left.val);
      }
      queue.push([node.left, min, node.val]);
    }

    if (right) {
      if (roots.has(right.val)) {
        node.right = roots.get(right.val);
        roots.delete(right.val);
      }
      queue.push([node.right, node.val, max]);
    }
  }

  if (roots.size === 1) return root;
  return null;
};
